test-app/tests/integration/components/headless-form-validation-test.gts

```diff
@@ -830,4 +845,778 @@ module('Integration Component HeadlessForm > Validation', function (hooks) {
+          '@validate is called with form data'
+        );
+
+        await fillIn('[data-test-first-name]', 'Tony');
+
+        assert.false(
+          validateCallback.calledTwice,
+          '@validate is not called while typing'
+        );
+
+        await blur('[data-test-first-name]');
+
+        assert.true(
+          validateCallback.calledTwice,
+          '@validate is called for revalidation on blur'
+        );
+        assert.true(
+          validateCallback
+            .getCall(1)
+            .calledWith({ ...data, firstName: 'Tony' }),
+          '@validate is called with form data'
+        );
+      });
+
+      test('field validation callback is called on blur', async function (assert) {
+        const data: TestFormData = { firstName: 'Tony', lastName: 'Ward' };
+        const validateCallback = sinon.spy();
+
+        await render(<template>
+          <HeadlessForm @data={{data}} @revalidateOn="blur" as |form|>
+            <form.field
+              @name="firstName"
+              @validate={{validateCallback}}
+              as |field|
+            >
+              <field.label>First Name</field.label>
+              <field.input data-test-first-name />
+            </form.field>
+            <form.field @name="lastName" as |field|>
+              <field.label>Last Name</field.label>
+              <field.input data-test-last-name />
+            </form.field>
+            <button type="submit" data-test-submit>Submit</button>
+          </HeadlessForm>
+        </template>);
+
+        await fillIn('[data-test-first-name]', 'Foo');
+
+        assert.false(
+          validateCallback.called,
+          '@validate is not called while typing'
+        );
+
+        await blur('[data-test-first-name]');
+
+        assert.false(
+          validateCallback.called,
+          '@validate is not called until submitting'
+        );
+
+        await click('[data-test-submit]');
+
+        assert.true(
+          validateCallback.calledOnce,
+          '@validate is called when submitting'
+        );
+        assert.true(
+          validateCallback.calledWith('Foo', 'firstName', {
+            ...data,
+            firstName: 'Foo',
+          }),
+          '@validate is called with form data'
+        );
+
+        await fillIn('[data-test-first-name]', 'Tony');
+
+        assert.false(
+          validateCallback.calledTwice,
+          '@validate is not called while typing'
+        );
+
+        await blur('[data-test-first-name]');
+
+        assert.true(
+          validateCallback.calledTwice,
+          '@validate is called for revalidation on blur'
+        );
+        assert.true(
+          validateCallback.getCall(1).calledWith('Tony', 'firstName', {
+            ...data,
+            firstName: 'Tony',
+          }),
+          '@validate is called with form data'
+        );
+      });
+
+      test('validation errors are exposed as field.errors on blur', async function (assert) {
```

Comment:
Suggestion: `'validation errors are exposed as field.errors on blur'` -> `'validation errors are exposed as field.errors on @revalidateOn="blur'`

Reply:
I would disagree here, because
* we have that wrapped in a `module()` that already makes clear that we are testing the `@revalidateOn="blur'` scenario
* the `on blur` refers to what the user is doing here, not what the dev is passing to the component